P3394 - C++26 Annotation

zebullax opened the pull request:
[P3394 - Annotations for Reflection](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2025/p3394r4.html)

**Onboard C++26 annotation into Clang.**

The design is inspired/taken from the [clang P2996 experimental compiler](https://github.com/bloomberg/clang-p2996) meaning that annotations leverage attributes machinery 
As such we add the description of a generic annotation attribute in attr.td used to tablegen most of the needed utilities (accessors, cstr, etc.). This attribute wraps over an APValue that holds the result of evaluating the annotation arg.
In non dependent case this evaluation is done as we go over parsing the annotation, in dependent case we leverage _attrTemplateInstantiate_.

zebullax committed:
Add annotation to attribute tablegen

Recognize annotation starting token in attr parsing

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Build logic around parsing of annotation inside list

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Add sema handler for annotation

Add parsing for annotation

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Fix the copy around of common info for attribute

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Fill the doc for ParseAnnotationSpecifier

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Apply formatter

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

tbaederr commented on clang/lib/Parse/ParseDeclCXX.cpp:
```diff
@@ -4681,16 +4703,21 @@ void Parser::ParseCXX11AttributeSpecifierInternal(ParsedAttributes &Attrs,
 
-  bool AttrParsed = false;
+  bool hasAttribute = false;
+  bool hasAnnotation = false;
```
```suggestion
  bool HasAttribute = false;
  bool HasAnnotation = false;
```

tbaederr commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -6409,6 +6409,59 @@ static void handleRequiresCapabilityAttr(Sema &S, Decl *D,
+        S.Diag(P.first, P.second);
+
+      return;
+    } else if (!CE->getType()->isStructuralType()) {
```
Don't need the `else` here.

zebullax committed:
Address review

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

Undo dumb edit

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax commented on clang/utils/TableGen/ClangAttrEmitter.cpp:
```diff
@@ -4091,13 +4091,15 @@ EmitClangAttrTemplateInstantiateHelper(ArrayRef<const Record *> Attrs,
     for (auto const &ai : Args)
       ai->writeTemplateInstantiation(OS);
 
```
Thanks, looking.

zebullax committed:
Verify we are not emitting hasAttribute for Annotation spelling

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Fix checks to allow annotation on declaration

Add smoke test

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Merge branch 'main' into annotation

erichkeane commented on clang/include/clang/Basic/DiagnosticParseKinds.td:
```diff
@@ -798,6 +798,10 @@ def err_cxx11_attribute_forbids_ellipsis : Error<
   "default scope specifier for attributes is incompatible with C++ standards "
   "before C++17">, InGroup<CXXPre17Compat>, DefaultIgnore;
+def warn_cxx26_compat_annotation : Warning<
+  "annotation is a C++26 extension">, InGroup<CXXPre26Compat>, DefaultIgnore;
```
```suggestion
  "annotations are a C++26 extension">, InGroup<CXXPre26Compat>, DefaultIgnore;
```

erichkeane commented on clang/utils/TableGen/ClangAttrEmitter.cpp:
```diff
@@ -3860,6 +3860,11 @@ void EmitClangAttrHasAttrImpl(const RecordKeeper &Records, raw_ostream &OS) {
       else if (Variety == "HLSLAnnotation")
         HLSLAnnotation.emplace_back(R, SI);
+      else if (Variety == "AS_Annotation") {
+        // We should not be code gening anything with a C++26
```
Yeah, we should probably diagnose this, an annotation having a spelling is ill-formed.  I do wonder if we should instead just omit the 'spellings' in Attr.td and allow that/diagnose on 'has spellings'.

erichkeane commented on clang/utils/TableGen/ClangAttrEmitter.cpp:
```diff
@@ -4091,13 +4100,15 @@ EmitClangAttrTemplateInstantiateHelper(ArrayRef<const Record *> Attrs,
-       << "    }\n";
+    OS << ");\n";
+    OS << R.getValueAsString("PostInstantiationStmts");
+    OS << "return Result;\n";
```
Can you share what is going on here a bit better?  It isn't clear what the intent is here without examples.

erichkeane commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -6409,6 +6409,60 @@ static void handleRequiresCapabilityAttr(Sema &S, Decl *D,
+
+    if (!CE->EvaluateAsConstantExpr(Result, S.Context, CEKind)) {
+      S.Diag(CE->getBeginLoc(), diag::err_attribute_argument_type)
+          << "C++26 annotation" << 4 << CE->getSourceRange();
```
These magic numbers need the comments to explain what they mean.

erichkeane commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -6409,6 +6409,60 @@ static void handleRequiresCapabilityAttr(Sema &S, Decl *D,
 
+static void handleCxx26AnnotationAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+  Expr *CE = AL.getArgAsExpr(0);
+  if (CE->isLValue()) {
```
Can you include the standardeeze here as to what exactly is happening/the rules of what should happen on the RHS here?

zwuis replied:
I think most checks of `handleCxx26AnnotationAttr` should be (maybe indirectly) delegated to `CheckTemplateArgument` in SemaTemplate.cpp . `meta::reflect_constant(E)` requires `E` as a valid template argument.

erichkeane commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -6409,6 +6409,60 @@ static void handleRequiresCapabilityAttr(Sema &S, Decl *D,
+
+      CE = CopyResult.get();
+    } else {
+      ExprResult RVExprResult = S.DefaultLvalueConversion(AL.getArgAsExpr(0));
```
Is there a reason the normal copy-init like above is incorrect here?

erichkeane commented on clang/lib/Parse/ParseDeclCXX.cpp:
```diff
@@ -4442,6 +4443,27 @@ static bool IsBuiltInOrStandardCXX11Attribute(IdentifierInfo *AttrName,
+    return;
+
+  IdentifierTable &IT = Actions.PP.getIdentifierTable();
+  IdentifierInfo &Placeholder = IT.get("__annotation_placeholder");
```
Hmm, this is a little strange unfortunately.  Can we try to modify `ParsedAttr` to accept one without a name?  We already have special cases in there for other attribute kinds, and annotations not having one seems sensible.

zebullax replied:
Thanks, I took a crack at it in https://github.com/llvm/llvm-project/pull/166287/commits/91095572881f9f3f646ef19c555415602dd173bd , is that what you had in mind ?

cor3ntin reviewed (changes requested):
just making sure we don't merge that that before #164692

zebullax committed:
Fix logic on mixed attribute and annotation

Improve documentation

Emit error if annotation syntax found in tablegen

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax commented on clang/lib/Sema/ParsedAttr.cpp:
```diff
@@ -107,6 +107,13 @@ const ParsedAttrInfo &ParsedAttrInfo::get(const AttributeCommonInfo &A) {
 
+  // If this is an annotation then return an appropriate ParsedAttrInfo.
+  static const ParsedAttrInfo AnnotationAttrInfo(
+      AttributeCommonInfo::AT_CXX26Annotation, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, {},
```
Fixed I think

zebullax commented on clang/lib/Parse/ParseDeclCXX.cpp:
```diff
@@ -4700,6 +4727,31 @@ void Parser::ParseCXX11AttributeSpecifierInternal(ParsedAttributes &Attrs,
+    //  - We must not mix with an attribute
+    if (Tok.is(tok::equal)) {
+      if (!getLangOpts().CPlusPlus26) {
+        Diag(Tok.getLocation(), diag::warn_cxx26_compat_annotation);
```
Made this an error by default

zebullax commented on clang/lib/Parse/ParseDeclCXX.cpp:
```diff
@@ -4700,6 +4727,31 @@ void Parser::ParseCXX11AttributeSpecifierInternal(ParsedAttributes &Attrs,
+        SkipUntil(tok::r_square, tok::colon, StopBeforeMatch);
+        continue;
+      }
+      if (HasAttribute) {
```
Thought I did, but my test case had a trailing comma... anyway fixed

zebullax commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -6409,6 +6409,60 @@ static void handleRequiresCapabilityAttr(Sema &S, Decl *D,
+      return;
+    }
+  }
+  auto *Annot = CXX26AnnotationAttr::Create(S.Context, CE, AL);
```
Those extra `create` are removed, we should be back to regular create facilities.

zebullax committed:
Apply formatter

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax commented:
I have addressed the quickest comments and the mishandled case on mixing attribute and annotation.
I will address the remaining questions as I free up some time on my end 🙇

zebullax committed:
Fix UT

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

tbaederr reviewed (commented):
You need to end your comments in a period (I'm just saying this before Aaron does).

zebullax committed:
Close comments with .

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

erichkeane commented on clang/test/SemaCXX/cxx26-annotation.cpp:
```diff
@@ -0,0 +1,35 @@
+};
+
+template<class T>
+  [[=T::type()]] void h(T t); // expected-error {{type 'char' cannot be used prior to '::' because it has no members}}
```
Appreciate the swap!  This works, thanks.

erichkeane reviewed (commented):
A release note might be a good idea here to show progress on reflection.

I think annotations should ALSO only work under the `freflection` flag for the short-term, but that isn't in place yet, so we might need to wait on that to merge this.

zebullax committed:
Add test covering deleted copy on record type for annotation

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Add more test on structural type

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Remove placeholder attr name for annotation

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax commented on clang/include/clang/Basic/Attr.td:
```diff
@@ -934,6 +937,48 @@ def AlwaysInline : DeclOrStmtAttr {
+  }];
+
+  let PostInstantiationStmts = [{
+    Expr::EvalResult V;
```
I believe there might be a more canonical way to achieve that via `Sema::InstantiateAttrs` , I'm giving this a try and that will remove this PostInstantiateStmts thingy

zebullax replied:
Gave it a try in https://github.com/llvm/llvm-project/pull/166287/commits/9ddf0a8b4a6190fa8ae14e6c4e58a13c9cb02e87
I assume (?) there are cases that will break and when I find those, we'll either fix it and keep this approach or revert to the more handcrafty `PostInstantiationStmts`

zebullax committed:
Leverage InstantiateAttrs to re-evaluate APValue

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Add draft of release notes

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Return const ref AP Value from annotation attr

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Improve comments

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Detect copy constructible type for expression passed to annotation

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

tbaederr commented on clang/lib/Sema/SemaDeclAttr.cpp:
```diff
@@ -6409,6 +6409,84 @@ static void handleRequiresCapabilityAttr(Sema &S, Decl *D,
+  // annotation.
+  // Among other requirements it means for we need to check they obey
+  // is_copy_construbtible.
+  if (const CXXRecordDecl *RD = CE->getType()->getAsCXXRecordDecl(); RD) {
```
```suggestion
  if (const CXXRecordDecl *RD = CE->getType()->getAsCXXRecordDecl()) {
```

zebullax committed:
Merge branch 'main' into annotation

zebullax committed:
Format

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Use proper function for diagnostic on deleted function

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Merge branch 'main' into annotation

shafik reviewed (commented):
We need the proposal number in the title, summary and Release note. The summary should also provide a link to the proposal. We should not be assuming reviewers know these things and we should be providing all the information needed to review this PR self contained in the PR, at minimum.

The main point is that the Title/Summary should provide all the information needed for a reviewer to review the PR w/o having to go searching for more information. Secondly, we want the PR to be accessible for as many reviewers as possible and providing this information allows non-WG21 folks to effectively review this PR. We need to do a better job here.

zebullax committed:
Add p number to release note

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Merge branch 'main' into annotation

zebullax committed:
Merge branch 'main' into annotation

zebullax committed:
Merge branch 'main' into annotation

zebullax committed:
Merge branch 'main' into annotation

zebullax committed:
Stash work on no annotation for alias RHS

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

Remove subjects

Regress the parsing of annotation after the declarator-id of a variable

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

Fix regression on parsing declaration specifier for a variable declaration

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Merge branch 'main' into annotation

zebullax committed:
Update clang/lib/Sema/SemaDeclAttr.cpp

Co-authored-by: Timm Baeder <tbaeder@redhat.com>

zebullax committed:
Address review

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Format

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Merge branch 'main' into annotation

zebullax committed:
Merge branch 'main' into annotation

zebullax committed:
Merge branch 'main' into annotation

shafik reviewed (commented):
Is there a godbolt setup I can test this against? 

nvm just realized I need to use `-freflection-latest` for this to work.

zebullax committed:
Add test for annotation in using

Signed-off-by: acassagnes <acassagnes@bloomberg.net>

zebullax committed:
Merge branch 'main' into annotation

zebullax committed:
Merge branch 'main' into annotation

zebullax committed:
Merge branch 'main' into annotation
